pkg/autodiscovery/listeners/kube_endpoints.go

```diff
@@ -0,0 +1,336 @@
+	if !ok {
+		log.Errorf("Expected an Endpoints type, got: %v", obj)
+		return
+	}
+	// Cast the old object, consider it an add on cast failure
+	castedOld, ok := old.(*v1.Endpoints)
+	if !ok {
+		log.Errorf("Expected an Endpoints type, got: %v", old)
+		l.createService(castedObj, false)
+		return
+	}
+	if l.endpointsDiffer(castedObj, castedOld) {
+		l.removeService(castedObj)
+		l.createService(castedObj, false)
+	}
+}
+
+// endpointsDiffer compares two endpoints to only go forward
+// when relevant fields are changed. This logic must be
+// updated if more fields are used.
+func (l *KubeEndpointsListener) endpointsDiffer(first, second *v1.Endpoints) bool {
+	// Quick exit if resversion did not change
+	if first.ResourceVersion == second.ResourceVersion {
+		return false
+	}
+	// AD annotations on the corresponding service
+	if l.isEndpointsAnnotated(first) != l.isEndpointsAnnotated(second) {
+		return true
+	}
+	// Subsets
+	return subsetsDiffer(first, second)
+}
+
+// subsetsDiffer detects if two Endpoints have different subsets.
+// The function is separated from endpointsDiffer to facilitate testing.
+func subsetsDiffer(first, second *v1.Endpoints) bool {
+	// Subsets length
+	if len(first.Subsets) != len(second.Subsets) {
+		return true
+	}
+	// Addresses and Ports
+	for i := range first.Subsets {
+		if len(first.Subsets[i].Addresses) != len(second.Subsets[i].Addresses) {
+			return true
+		}
+		if len(first.Subsets[i].Ports) != len(second.Subsets[i].Ports) {
+			return true
+		}
+		// Addresses
+		for j := range first.Subsets[i].Addresses {
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
+				return true
+			}
+		}
+		// Ports
+		for j := range first.Subsets[i].Ports {
+			if first.Subsets[i].Ports[j].Port != second.Subsets[i].Ports[j].Port {
+				return true
+			}
+			if first.Subsets[i].Ports[j].Name != second.Subsets[i].Ports[j].Name {
+				return true
+			}
+		}
+	}
+	// No relevant change
+	return false
+}
+
+// isEndpointsAnnotated looks for the corresponding service of a kubernetes endpoints object
+// and returns true if the service has endpoints annotations, otherwise returns false.
+func (l *KubeEndpointsListener) isEndpointsAnnotated(kep *v1.Endpoints) bool {
+	ksvc, err := l.serviceLister.Services(kep.Namespace).Get(kep.Name)
+	if err != nil {
+		log.Debugf("Cannot get Kubernetes service: %s", err)
+	}
+	if ksvc != nil {
+		if _, found := ksvc.Annotations[kubeEndpointsAnnotationFormat]; found {
+			return true
+		}
+	}
+	return false
+}
+
+func (l *KubeEndpointsListener) createService(kep *v1.Endpoints, firstRun bool) {
+	if kep == nil {
+		return
+	}
+	if !l.isEndpointsAnnotated(kep) {
+		// Ignore endpoints with no AD annotation on their corresponding service
+		return
+	}
+
+	eps := processEndpoints(kep, firstRun)
+
+	l.m.Lock()
+	l.endpoints[kep.UID] = eps
+	l.m.Unlock()
+
```

Comment:
I wanted to use the same approach/style we use in the Service listener `kube_services.go`, we don't log anything there when creating a service.
But I do agree with you that we should log something 👍(in Debug imo, because all our logs in kubelet, service, ecs, docker listeners are in Debug)